src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs

```diff
@@ -587,36 +584,64 @@
             return result.ToImmutableAndFree();
         }
 
+        private static SynthesizedLocalKind SpillSynthesizedLocalKind(BoundExpression expression)
+        {
+            switch (expression.Kind)
+            {
+                case SpillSequenceBuilderKind:
+                    return SynthesizedLocalKind.Spill;
+                case BoundKind.LoweredIsPatternExpression:
+                    return SynthesizedLocalKind.LoweringTemp;
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(expression.Kind);
+            }
+        }
+
+        private static bool ShouldSpill(BoundExpression expression)
+        {
+            switch (expression.Kind)
+            {
+                case SpillSequenceBuilderKind:
+                case BoundKind.LoweredIsPatternExpression:
```

Comment:
>BoundKind.LoweredIsPatternExpression [](http://example.com/codeflow?start=21&length=36)

I think we should avoid triggering spilling just because we run into LoweredIsPatternExpression. The emit layer is able to handle the node, therefore, the node can be left "unchanged" (i.e. just its parts visited and reassembled back into LoweredIsPatternExpression) during a visit.

Reply:
>I think we should avoid triggering spilling just because we run into LoweredIsPatternExpression

The core problem we're solving with this is that patterns are lowered as statements, hence the requirement for empty stack. (This is the same issue mentioned in [block expressions](https://github.com/dotnet/csharplang/discussions/1167) proposal, which supposedly is going to use a SpillSequence.)

We can't always rewrite the dag into an expression, for instance, there's BoundSwitchDispatch which could be emitted as a range check but we don't know that this until emitting.

Reply:
> The core problem we're solving with this is that patterns are lowered as statements

Not any longer, the LoweredIsPatternExpression is an expression. Not a trivial expression, but an expression. And we do understand how to deal with it all the way to the emit layer.

> hence the requirement for empty stack. 

I do not think there is a requirement like that for the new node. It really shouldn't matter what is on the stack. When we emit the new node, it just pushes a value to the stack. Obviously, we avoid the push when we emit conditional branch, but in that case LoweredIsPatternExpression has no impact on the stack

Reply:
>It really shouldn't matter what is on the stack. 

To be clear, EmitStatement asserts that stack is empty, do you mean we bypass this and just emit the code as-is?

Also, looks like there's no requirement for an empty stack for `switch` in ecma spec §III.3.66? If so I think we can just emit the whole thing regardless of the state of eval stack.

Reply:
I was wrong about this for so long. https://sharplab.io/#v2:DYLgZgzgNALiCWwoBMQGoA+BYAUL/OwApjAARikAepAnqQF6kC8puppARB6WqQLYBDGAGMAFg1IB3eDHEZSARlIBaAHycYAJwCuRbvIBMK9RzADgEPW3Y2gA

Reply:
> To be clear, EmitStatement asserts that stack is empty, do you mean we bypass this and just emit the code as-is?

In context of the new node that assert doesn't make sense. What makes sense to assert, is that the stack is at a certain level. By default, that level is 0, but in context of the new node, that level is probably the size of the stack at which we reached the expression. Quite possibly we can simply adjust the assert to no longer use a constant for the expected size of the stack.